Prometheus: Add a keyboard shortcut to toggle all exemplars

bobrik opened the pull request:
Sometimes it's hard to see the quantile line behind all the exemplars (far from the worst example):

<img width="1362" alt="image" src="https://user-images.githubusercontent.com/89186/223927519-4e61a16a-598f-40bd-9354-dc854de5cbb6.png">

It's next to impossible to find the line you're looking for and it's too many clicks to turn off exemplars.

This commit adds a `d x` keyboard shortcut to toggle exemplar visibility on all prometheus queries. Unlike with legends, the logic is simpler and it does a pure toggle as opposed to "majority toggle" as with legends.

Since exemplars might not be loaded, this also requires refreshing the data. For the same reason, toggling a single panel is not supported, as it will make the panel data out of sync with the rest of the dashboard.

I'm happy to add tests and whatever else is missing if this makes sense and somebody can point me to examples.

bobrik committed:
Add a keyboard shortcut to toggle all exemplars

Sometimes it's hard to see the quantile line behind all the exemplars.

This commit adds a `d x` keyboard shortcut to toggle exemplar visibility
on all prometheus queries. Unlike with legends, the logic is simpler
and it does a pure toggle as opposed to "majority toggle" as with legends.

Since exemplars might not be loaded, this also requires refreshing the data.
For the same reason, toggling a single panel is not supported, as it will
make the panel data out of sync with the rest of the dashboard.

bobrik commented on public/app/features/dashboard/state/DashboardModel.ts:
```diff
@@ -1094,6 +1094,20 @@ export class DashboardModel implements TimeModel {
+  toggleExemplarsForAll() {
+    for (const panel of this.panels) {
+      for (const target of panel.targets) {
+        if (target.datasource.type !== 'prometheus') {
```
Apparently rows are panels too and they have queries, so this fails for them.

bobrik commented:
Status update: this PR has been ignored.

leeoniya commented:
thanks for the PR @bobrik, sorry about the delay. this one fell through the cracks. i've asked the metrics team to take a look. cc @gtk-grafana

gtk-grafana commented:
@bobrik Thanks for taking a look at this! It looks like we've got a lint issue on the frontend which needs to be fixed before merge, but I'm also seeing a bug in which the "x" key toggle which takes people to explore overrides the "d + x" key command added in this PR, so instead of removing the exemplars when typing "d + x", I'm sometimes (about every other time) moved to the explore page, this makes for a pretty frustrating UX.

gtk-grafana commented:
I'm reaching out to the dashboard squad and I'll let you know what they think, maybe we can use a different set of keys that doesn't conflict with existing commands.

ivanortegaalba commented:
> @bobrik Thanks for taking a look at this! It looks like we've got a lint issue on the frontend which needs to be fixed before merge, but I'm also seeing a bug in which the "x" key toggle which takes people to explore overrides the "d + x" key command added in this PR, so instead of removing the exemplars when typing "d + x", I'm sometimes (about every other time) moved to the explore page.

From the dashboard's perspective, it is ok to use `x`. I cannot find any other command that makes sense. It is also namespaces by `d` so this can happen with other shortcuts as well. Actually, I don't know why `x` redirects to explore. It should be `g + x`.

gtk-grafana commented:
Looks like the 'x' shortcut comes from here: https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/utils/getPanelMenu.ts#L154.
I think we can just remove that shortcut line and this PR is good to go?
@bobrik @ivanortegaalba

ivanortegaalba commented:
> Looks like the 'x' shortcut comes from here: https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/utils/getPanelMenu.ts#L154.
> 
> I think we can just remove that shortcut line and this PR is good to go?
> 
> @bobrik @ivanortegaalba 

I agree. That shortcut should only work when a panel is selected, and it should be combined with `p x`. So we can remove that shortcut and merge this one.

bobrik committed:
Use "p x" to navigate to panel explore rather than just "x"

It's more consistent with other panel level shortcuts.

It also doesn't conflict with the global "x" to toggle exemplars that way.

bobrik commented:
I replaced the `x` shortcut with `p x` for consistency, made the `typecheck` lint happy, and rebased.

I'm not sure what `betterer` is unhappy about here:

```
・ New issue in
"/drone/src/public/app/features/dashboard/state/DashboardModel.ts"!
・
・   /drone/src/public/app/features/dashboard/state/DashboardModel.ts
・ > 1 |
・     | ^ Unexpected any. Specify a different type.
```

There is no `any` on line 1 there.

gtk-grafana commented on public/app/features/dashboard/state/DashboardModel.ts:
```diff
@@ -1156,6 +1157,21 @@ export class DashboardModel implements TimeModel {
+          continue;
+        }
+
+        const promTarget = target as PromQuery;
```
This is the assertion that is failing the betterer check in drone, you can run yarn betterer -u and commit the results to get the tests passing.

bobrik replied:
Thanks, drone is happy now.

bobrik commented on public/app/core/services/keybindingSrv.ts:
```diff
@@ -259,7 +259,7 @@ export class KeybindingSrv {
     // jump to explore if permissions allow
     if (contextSrv.hasAccessToExplore()) {
-      this.bindWithPanelId('x', async (panelId) => {
+      this.bindWithPanelId('p x', async (panelId) => {
```
Here's how it should work:

* `p x` goes to explore (a panel needs to be in focus)
* `d x` toggles exemplars

Are you seeing something different?

gtk-grafana replied:
Nope this is what I'm seeing as well, I'm used to using `g e` for going to global - explore, but `p x` works too

gtk-grafana reviewed (approved):
Looks good, thanks for all your hard work on this!

bobrik commented:
Thanks @gtk-grafana!
